### PR TITLE
Feature/project 4

### PR DESCRIPTION
**Pull request by RicardoChapaRomero**
### What does this PR do?
It provides a solution to week 4's calendar coding challenge. It implements an algorithmic solution to schedule available meeting times between optional and required attendees. The complexity of the algorithm is O(MxN) in runtime and O(N) on space used. 

### Why is this PR necessary? 
It concludes the starter projects of this 2020 STEP internship, focusing mostly on Unit Testing with TDD. This PR can serve as guidance when it comes unit testing on our 2020 STEP Capstone Project.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -12,12 +12,243 @@
...
+
+    //Get the times where the attendees are not free
+    final PriorityQueue<TimeRange> requiredAttendeesOccupiedTimes = 
+      getAttendeesOccupiedTimes(events, meetingAttendees, true);
+    final PriorityQueue<TimeRange> optionalAttendeesTimeList = 
+      getAttendeesOccupiedTimes(events, meetingAttendees, false);
+
+    /** 
+    * Look for time ranges for optional attendees 
+    * if required attendees all available all day
```
Comment (akihiroota87) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
    * if required attendees are available all day 
```
Nit: Is this a typo?

Comment (akihiroota87) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Also, I think the /** */ style of comments is used for JSDoc. If you have a multiline comment, then I think it's okay to use //. So like below:

// Look for time ranges for optional attendees
// if required attendees are available all day

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -12,12 +12,243 @@
...
+      }
+    }
+
+    else if (requiredAttendeesOccupiedTimes.size() == 1) {
+      return 
+        getTimeRangesFromSingleEvent(requiredAttendeesOccupiedTimes, meetingDuration);
+    }
+
+    /**
+    * Convert all overlapping events in to compressed events for faster processing
```
Comment (akihiroota87) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Same here, comment can be written using //.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -12,12 +12,243 @@
...
+    }
+
+    meetingTimesForRequiredAttendees = 
+      getAvailableTimeRanges(compressedEvents, meetingDuration);
+
+    Collection<TimeRange> meetingTimesForAllAttendees = 
+      meetingTimesForRequiredAttendees;
+
+    /**
+    * If the optional attendees have occupied times then
```
Comment (akihiroota87) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Same here

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -12,12 +12,243 @@
...
+          if (event.next().equals(optionalTimeRange)) {
+            event.remove();
+            break;
+          }
+        }
+      }
+    }
+
+    /**
+    * If there's no room for optional attendees then
```
Comment (akihiroota87) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Same here

**Review by akihiroota87 — APPROVED** [review]
Looks good! Just some small things related to comments.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -12,12 +12,243 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+/**
```
Comment (mschillaci) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: This should be right above the line "public final class FindMeetingQuery {"

**Review by mschillaci — APPROVED** [review]
LGTM

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (RicardoChapaRomero) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Thanks for pointing it out Mark!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (RicardoChapaRomero) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Thanks for the recommendation Aki!

**Pull request merged**